File: elixir/isbn-verifier/lib/isbn_verifier.ex
defmodule IsbnVerifier do
  @doc """
    Checks if a string is a valid ISBN-10 identifier

    ## Examples

      iex> IsbnVerifier.isbn?("3-598-21507-X")
      true

      iex> IsbnVerifier.isbn?("3-598-2K507-0")
      false

  """
  @spec isbn?(String.t()) :: boolean
  def isbn?(isbn) do
    with [isbn] <- Regex.run(~r/^\p{N}{9}[\p{N}X]$/, String.replace(isbn, "-", "")) do
      isbn
      |> to_charlist()
      |> Enum.map(fn
        d when d in ?0..?9 -> d - ?0
        ?X -> 10
      end)
      |> Enum.with_index(fn d, i -> d * (10 - i) end)
      |> Enum.sum()
      |> rem(11) == 0
    end
  end
end

File: zig/isbn-verifier/isbn_verifier.zig
pub fn isValidIsbn10(s: []const u8) bool {
    var sum: i64 = 0;
    var c: i64 = 10;
    for (s) |letter| {
        if (letter >= '0' and letter <= '9') {
            sum += (letter - '0') * c;
            c -= 1;
        } else if (c == 1 and letter == 'X') {
            sum += 10;
            c -= 1;
        } else if (letter != ' ' and letter != '-') {
            return false;
        }
    }
    return c == 0 and @rem(sum, 11) == 0;
}
